### DISCUSSION TITLE
Changing text size for the REPL

### DISCUSSION THREAD

**Discussion by andypiper** [shorepine/tulipcc#143] [Ideas]
I'm currently using the T-Deck port.

It would be pretty helpful if I could have the REPL use a font that would provide at least a little more text density for the screen. I note that there's a `fonts.py` example.

Where would I look to modify the code for the T-Deck port to use one of the other typeface options in the REPL - or is that non-trivial?

**Comment by bwhitman** [comment]
The REPL font is currently fixed to an 8x12 pixel font defined in [`smallfont.c`](https://github.com/bwhitman/tulipcc/blob/main/tulip/shared/smallfont.c). I woudl _love_ to make smallfont something that someone can change in Micropython at runtime, but for now its' a fixed const header of bytes. You can change it manually if you're ok with compiling Tulip on your own. But wouldn't it be cool to have a `tulip.repl_font('new_font.fon')` command? 

Other than changing the font data per character in ASCII, we'd also want to change the FONT_HEIGHT and FONT_WIDTH defines at runtime.

**Comment by andypiper** [comment]
It seems as though we'd want to use the u8 fonts in the REPL rather than hard-coding the existing small font, maybe - and beyond that, it would be super cool to be able to change the font at runtime, yes. I know Pimoroni are doing some interesting work in this space on top of MicroPython with their [new PicoVector library](https://github.com/pimoroni/pimoroni-pico/tree/main/libraries/pico_vector) and support for Alright Fonts, but I don't think the performance is too great yet.

Thanks, I might look into building with a more dense font specifically for the T-Deck, although I also imagine I'll get interested in the desktop and actual Tulip with screen at some stage...

**Reply by bwhitman** [reply]
Sure thing. One thing to keep in mind is that to keep high REPL / editor text performance on the esp32s3, we use an 8px unrolled loop in the scanline drawing GDMA process that reads the font data as it draws. So far it's the fastest text drawing thing I've seen on an ESP at our high resolutions. This will matter less for the T-Deck as you're only drawing to 1/8th of the pixels as Tulip CC. 

Any font you use will have to be arranged in the same orientation as `smallfont.c` and if it's narrower than 8px you'll have to change the unrolling. We profiled a for loop there at almost 2x slower than the unrolled one on esp32s3. It's all here: 
https://github.com/bwhitman/tulipcc/blob/cf542703bc3f091f09d1de89d5515b169a50940a/tulip/shared/display.c#L224

**Comment by bwhitman** [comment]
For those not on Discord, we just added support for two REPL fonts -- the default 8x12 one, and a smaller 6x8 one. It's a compile time option for all versions of Tulip. See the flashing / desktop / t-deck pages for more detail on how to apply it. (Small is default on T-Deck.) 

If you want to load in your own fonts: what i did was load a ttf from https://int10h.org/oldschool-pc-fonts/fontlist/font?portfolio_6x8 into this app https://lvgl.io/tools/font_conv_v5_3 , loaded that into `smallfont.c`, and then made changes to `FONT_HEIGHT` and `FONT_WIDTH` and put some #ifdefs in `display.c`: https://github.com/bwhitman/tulipcc/compare/cf542703bc3f...c0dcb1da31a3
